feat!: Mark transactions as reverted on L1

just-mitch opened the pull request:
# Introduction

Presently, if a transaction is reverted, the base rollup does not mark it as reverted: it merely has its revertible side effects dropped. This issue will update the base rollup to mark reverted transactions.

# Solution Design

## New `RevertCode` in `TxEffect`

Create a new class in `circuits.js` to hold a `RevertCode`. It will be a safe wrapper around an `Fr` that will be used to store the revert code of a transaction.

`0` will indicate success, and any other value will indicate failure. Presently, only `1` is used to indicate general failure, but we'll leave the door open for future expansion.

`TxEffect` will be updated to include a `RevertCode`.

### Ethereum `status`

Ethereum stores their `status` in a `uint64`, but only currently use `0` to indicate failure and `1` for success. [More info](https://github.com/ethereum/EIPs/issues/98) (Thanks @spalladino !)

### Size considerations

We'll eventually want to add `gasUsed` and `gasPrice` to `TxEffect`. Using a full field is wasteful, but using a smaller size will make the encoding and hashing more complex (see below). We'll implement with a full field and have a stacked PR to optimize the size.

## Kernel Constraints

The `RevertCode` will become part of the content commitment for a transaction.

The content commitment is computed by the base rollup in `compute_tx_effects_hash`, which:

- accepts a `CombinedAccumulatedData`
- operates over `Field`s

We will therefore update the `CombinedAccumulatedData` to include the `RevertCode`.

A fallout from that is we will move the current `reverted` flag from `PublicKernelCircuitPublicInputs` to be part of:

- `PrivateAccumulatedNonRevertibleData`
- `PublicAccumulatedNonRevertibleData`

This is because we:

1. build up a `CombinedAccumulatedData` during private execution
2. split them into `PrivateAccumulatedNonRevertibleData` and `PrivateAccumulatedRevertibleData` in the private kernel tail
3. convert those `PrivateAccumulated...` into `PublicAccumulated...` during public kernel execution as needed
4. recombine them back into `CombinedAccumulatedData` before feeding back into base rollup

## Encoding

The `RevertCode` will come first in the encoded `TxEffect`.

That is we will publish:

```
|| 32 bytes for RevertCode || ... Existing PublishedTxEffect ... ||
```

## L1 Tx Decoder

We'll need to update the availability oracle to compensate for the new flag.

## `TxReceipt`

We'll add a new `TxStatus` that is `REVERTED = 'reverted'`.

We need to update `getSettledTxReceipt` to inspect the `TxEffect` and set the `status` accordingly.

# Test Plan

## Unit Tests

- Serde of `TxEffect` in TS and solidity.

## E2E Tests

- Add checks for tx statuses of `REVERTED` in the `e2e_fees`, `e2e_deploy_contract`, and `dapp_testing`.

# Documentation Plan

Will start a page in the "Data publication and availability" section of the yellow paper describing the format of the block submitted.

# Especially Relevant Parties

- @PhilWindle
- @benesjan
- @alexghr
- @LeilaWang
- @LHerskind

# Plan Approvals

**_👋 Please add a +1 or comment to express your approval or disapproval of the plan above._**

spalladino commented:
> I propose to rename TxEffect to PublishedTxEffect, and add two new classes:
> 
> PublishedTxMetadata which will contain the published metadata for a transaction
> PublishedTx which will contain the PublishedTxEffect and PublishedTxMetadata

Why not just include the revert flag as part of the TxEffect? It's data that gets published on chain and that needs to be passed into the rollup circuits, just like any other tx effect. I think I'm not seeing why it's different from a public write, for instance.

Aside from that, +1 from me!

just-mitch commented:
@spalladino Totally can all be in the same place. The separation was just for more clear semantics in TS: `Effect` holds what was "done", and `Metadata` holds information about how it was done.

I didn't like putting the `reverted` flag in the `Effects` because it isn't a "write", or what I think of as a side effect.

So maybe the solution is to put them all together, but just rename `TxEffect` to `PublishedTx`?

benesjan commented:
@just-mitch I personally really like the name `TxEffect` because it really just contains the effects (nullifiers and such). When something is named a Tx I imagine that it's something which contains calldata and a signature (or proof in our case). For this reason I would prefer just including the revert flag in the `TxEffect` as Palla proposes.

PhilWindle commented:
>Note that the reverted flag comes from the PublicKernelCircuitPublicInputs. If we're coming straight from private, we'll manually set the flag to false now, but in the future we'll probably need a specialized circuit to handle the case where there is no public component. @LeilaWang ?

Can we not add it to the private kernel public inputs with value always `false`? It then gets propagated through the public kernels and set to false in case of revert.

PhilWindle commented:
I cna understand the motivation with splitting things out. As other say though I suspect it's simpler to keep as is.

just-mitch commented:
@benesjan If the revert flag is set, then all *revertible* side effects have been stripped out, but non-revertible side effects will be present in `TxEffect`. There is some info on this [here](https://yp-aztec.netlify.app/docs/gas-and-fees/fee-payments-and-metering#transaction-phases), but it does demand more documentation depth. 

@PhilWindle I think we will need a separate circuit for private in the event when there is no public execution regardless because the base rollup expects recombined non/revertible (#5013). So I can imagine two tails for private- one that splits non/revertible (as we currently have), and another that says  "oh there's nothing to split, I'll reformat this for the base rollup". I expect we'd choose in typescript after executing all required init/inner private kernels which tail to choose.

Regardless, I think I've also convinced myself that it is okay to put the flag in `TxEffect`: I had to switch it in my mind to read as "the result of this tx was these nullifiers etc *and* a raised `reverted` flag, *and* `consumedGas` etc", instead of my original conception of "the result of this tx was these nullifiers etc; the *attributes* of this tx include a raised `reverted` flag". Unless someone else comes along in the next few hours with a strong favor to the split out metadata I will change the original comment to reflect. Thanks all!

just-mitch commented:
Okay, updated the design above to reflect the conversations.

Also, added a section on documentation.

Also, updated the L1 encoding to reflect that we will need to publish the `reverted` flag as a `byte`, not a `bit` (since we cannot serialize to a bit in typescript)

just-mitch committed:
add revertcode to block

just-mitch committed:
update block store to return reverted status

just-mitch committed:
documentation updates

just-mitch commented on noir-projects/noir-protocol-circuits/crates/rollup-lib/src/components.nr:
```diff
@@ -146,6 +146,7 @@ pub fn compute_tx_effects_hash(combined: CombinedAccumulatedData) -> [Field; NUM
     // 1 unencrypted logs hash --> 1 sha256 hash -> 32 bytes  -> 2 fields | Beware when populating bytes that it is only 32!
     let mut txs_effects_hash_input = [0; TX_EFFECTS_HASH_INPUT_FIELDS];
 
+    let reverted = combined.reverted;
```
This was the primary impetus for me simply adding the `reverted` code to `CombinedAccumulatedData`. This way we get the flag as an out from the private circuits as well.

just-mitch commented on noir-projects/noir-protocol-circuits/crates/types/src/abis/kernel_circuit_public_inputs/public_kernel_circuit_public_inputs.nr:
```diff
@@ -13,5 +13,4 @@ struct PublicKernelCircuitPublicInputs {
     needs_setup: bool,
     needs_app_logic: bool,
     needs_teardown: bool,
-    reverted: bool,
```
This got pushed input PublicAccumulatedNonRevertibleData.

just-mitch commented on noir-projects/noir-protocol-circuits/crates/types/src/abis/public_circuit_public_inputs.nr:
```diff
@@ -48,7 +48,7 @@ struct PublicCircuitPublicInputs{
     prover_address: AztecAddress,
 
-    reverted: bool,
+    reverted: Field,
```
Because the base rollup computes its content commitment based on accumulating `Field`s, I felt this was appropriate for a first implementation. Ideally we will trim down the size, but that will require specialization of `compute_tx_effects_hash`

just-mitch commented on yarn-project/circuits.js/src/structs/revert_code.ts:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Wrapper class over a field to safely represent a revert code.
+ */
+export class RevertCode {
```
I felt this was safer than exposing a raw `Fr`.

alexghr replied:
I like the parallel to unix's exit code, 0 for success != 0 for failure!

just-mitch committed:
final cleanup

just-mitch commented:
Built to spec, barring the caveat that we publish a full `Fr` for the `reverted` flag. See comments above for my reasoning. I suspect we can/will optimize this in the future.

just-mitch committed:
fix sidebar

PhilWindle commented on noir-projects/noir-protocol-circuits/crates/public-kernel-lib/src/common.nr:
```diff
@@ -36,15 +36,19 @@ pub fn validate_inputs(public_call: PublicCallData) {
     circuit_outputs: &mut PublicKernelCircuitPublicInputsBuilder
 ) {
-    circuit_outputs.reverted = previous_kernel.public_inputs.reverted | public_call.call_stack_item.public_inputs.reverted;
+    if previous_kernel.public_inputs.end_non_revertible.reverted != 0 {
```
So, we always take the revert code of the previous kernel if non-zero. If zero we take the result of this iteration.

Sounds correct.

PhilWindle commented on yarn-project/circuit-types/src/tx/tx_receipt.ts:
```diff
@@ -11,6 +11,7 @@ export enum TxStatus {
   DROPPED = 'dropped',
   MINED = 'mined',
```
Not for this PR but I can't help but think we should rename 'MINED'. Reverted transactions are also 'mined'.

just-mitch replied:
Very true. I think they should roughly correspond with the RevertCode options. I added a TODO on https://github.com/AztecProtocol/aztec-packages/issues/4096

PhilWindle reviewed (commented):
This all looks ok. I think @LeilaWang should probably look over it.

Are we using a full field for the revert code in anticipation of encoding more data into it?

alexghr commented on yarn-project/update-snapshots.sh:
:heart:

just-mitch commented:
* **#5309** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/5309?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5301** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/5301?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5226** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/5226?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @just-mitch and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

just-mitch commented:
### Merge activity

* **Mar 20, 1:08 PM EDT**: @just-mitch started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/5226).
